app/controllers/auth/AuthAction.scala

```diff
@@ -0,0 +1,100 @@
+import play.api.i18n.Messages
+import play.api.mvc.{AnyContent, Request, Result}
+import services.DelegationService
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{ExecutionContext, Future}
+
+trait AuthAction extends AtedBaseController with AuthorisedFunctions {
+
+  def unauthorisedUrl(isSa: Boolean = false): Result = Redirect(controllers.routes.ApplicationController.unauthorised(isSa).url)
+
+  val delegationService: DelegationService
+
+  def authorisedForNoEnrolments(body: StandardAuthRetrievals => Future[Result])
+                                (implicit request: Request[AnyContent],
+                  ex: ExecutionContext,
+                  hc: HeaderCarrier,
+                  messages: Messages): Future[Result] = {
+    authorised(
+      AffinityGroup.Organisation or AffinityGroup.Agent
+    ).retrieve(
+      Retrievals.allEnrolments and Retrievals.affinityGroup and Retrievals.internalId
+    ){
+      case enrolments ~ affinityGroup ~ Some(_) =>
+          body(StandardAuthRetrievals(enrolments.enrolments, affinityGroup, None))
+    } recover {
+      case e: AuthorisationException => handleException(e)
+    }
+  }
+
+  def validateAgainstSaEnrolment(enrolments: Enrolments): Boolean = {
+
+    enrolments.enrolments.count{ enrolment =>
+      enrolment.key match {
+        case "HMRC-ATED-ORG" | "HMRC-AGENT-AGENT" | "IR-SA" => true
+        case _                                              => false
+      }
+    } == 1 && enrolments.getEnrolment("IR-SA").isDefined
+  }
+
+  def authorisedAction(body: StandardAuthRetrievals => Future[Result])
+                      (implicit request: Request[AnyContent],
+                       ex: ExecutionContext,
+                       hc: HeaderCarrier,
+                       messages: Messages): Future[Result] = {
+    authorised(
+      (Enrolment("HMRC-ATED-ORG") or Enrolment("HMRC-AGENT-AGENT") or Enrolment("IR-SA")) and
+      ConfidenceLevel.L50 and
+      (AffinityGroup.Organisation or AffinityGroup.Agent)
+    ).retrieve(
+      Retrievals.allEnrolments and Retrievals.affinityGroup and Retrievals.internalId
+    ){
+      case enrolments ~ affinityGroup ~ Some(internalId) =>
+        if (validateAgainstSaEnrolment(enrolments)) {
+          Future.successful(unauthorisedUrl(true))
+        } else {
+          delegationService.delegationCall(internalId) flatMap { delegationModel =>
+            body(StandardAuthRetrievals(enrolments.enrolments, affinityGroup, delegationModel))
+          }
+        }
+    } recover {
+      case e: AuthorisationException => handleException(e)
+    }
+  }
+
+  def handleException(ae: AuthorisationException): Result  = ae match {
+    case nas: NoActiveSession =>
+      Logger.debug("DefaultAuthAction:Refine - NoActiveSession:" + nas)
```

Comment:
AMBER - Unsure about this one but maybe we want to log this at WARN?

Reply:
this applies to the AuthorisationException below too.

Reply:
Definitely not a WARN, definitely not a DEBUG

Reply:
Done

Reply:
Amended to level info